### PR TITLE
fix(platform/gitlab): resolve assignees/reviewers from groups

### LINKED ISSUE #14777: Gitlab: Support for reviewers taken from groups used in CODEOWNERS file
### What would you like Renovate to be able to do?

We use Gitlab groups as nested organisational structure, teams are assigned as code owners. It would be great if MR reviewers could be set using the same approach. Currently with `reviewersFromCodeOwners=true` and `composer.lock @team/composer`-like rules in `CODEOWNERS` file the result of Renovate pass is:

```
 INFO: PR created (repository=foo/bar, branch=feature/renovate/aws-aws-sdk-php-3.x-lockfile)
       "pr": 7922,
       "prTitle": "fix(composer): update dependency aws/aws-sdk-php to v3.215.1"
 WARN: Failed to add assignees (repository=foo/bar, branch=feature/renovate/aws-aws-sdk-php-3.x-lockfile)
       "iid": 7922,
       "assignees": ["team/composer"]
 WARN: Failed to get IDs of the new reviewers (repository=foo/bar, branch=feature/renovate/aws-aws-sdk-php-3.x-lockfile)
       "err": {
         "message": "Cannot read property 'id' of undefined",
         "stack": "TypeError: Cannot read property 'id' of undefined\n    at getUserID (/usr/src/app/node_modules/renovate/lib/modules/platform/gitlab/http.ts:10:13)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at /usr/src/app/node_modules/p-map/index.js:57:22"
       }
```

Script that determines users from `CODEOWNERS` should fallback to group check if user ID can't be resolved from rule entry. Most probably [Members API](https://docs.gitlab.com/ee/api/members.html#list-all-members-of-a-group-or-project) should be used for that.

### If you have any ideas on how this should be implemented, please tell us here.

For each item in `CODEOWNERS` rule entry (e.g. `@john`, `@admins`, `@team/devops`):

- check if it's user
  - if yes, use ID and set as reviewer
  - if not, check if it's a group
    - if yes, retrieve all members of a group and use their IDs for setting as reviewers
    - if not, print error and ignore reviewers' assignment

PS. I choose "Maybe" because I really would like to dig into the code and try to implement it, but I don't know if I can do it (time, knowledge).

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by justinas-b**
## Changes

Adding ability to resolve assignees/reviewers in GitLab if groups are used in CODEOWNERS instead of users

## Context

This pull request is to address https://github.com/renovatebot/renovate/issues/14777 issue and is based on original pull request (https://github.com/renovatebot/renovate/pull/15835) from @suprememoocow

- closes #14777

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by justinas-b** [comment]
It seems build is failing due to https://stackoverflow.com/questions/62324458/typescript-error-property-flat-does-not-exist-on-type-string-unknown

Should i update  [tsconfig.spec.json](https://github.com/renovatebot/renovate/blob/main/tsconfig.spec.json)?

**Comment by viceice** [comment]
> It seems build is failing due to https://stackoverflow.com/questions/62324458/typescript-error-property-flat-does-not-exist-on-type-string-unknown
> 
> Should i update  [tsconfig.spec.json](https://github.com/renovatebot/renovate/blob/main/tsconfig.spec.json)?

nope, there's something wrong with your code. there is no method flat on type promise. 😉

File: lib/modules/platform/gitlab/index.ts [hunk 1]
```diff
@@ -1013,12 +1013,25 @@ export async function addReviewers(
   // Gather the IDs for all the reviewers we want to add
   let newReviewerIDs: number[];
   try {
-    newReviewerIDs = await p.all(newReviewers.map((r) => () => getUserID(r)));
+    newReviewerIDs = await p.all(
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1016]:
```suggestion
    newReviewerIDs = (await p.all(
```

File: lib/modules/platform/gitlab/index.ts [hunk 2]
```diff
@@ -1013,12 +1013,25 @@ export async function addReviewers(
...
+      newReviewers.map((r) => async () => {
+        try {
+          return [await getUserID(r)]
+        } catch (err) {
+          // Unable to fetch userId, try resolve as a group
+          return getMemberUserIDs(r)
+        }
+      }),
+    ).flat();
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1025]:
```suggestion
    )).flat();
```

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
Needs test fixes

File: lib/modules/platform/gitlab/index.ts [hunk 3]
```diff
@@ -50,7 +50,7 @@ import type {
 } from '../types';
 import { repoFingerprint } from '../util';
 import { smartTruncate } from '../utils/pr-body';
-import { getUserID, gitlabApi, isUserBusy } from './http';
+import { getUserID, getMemberUserIDs, gitlabApi, isUserBusy } from './http';
```
Comment (justinas-b) [inline lib/modules/platform/gitlab/index.ts:53]:
```suggestion
import { getMemberUserIDs, getUserID, gitlabApi, isUserBusy } from './http';
```

**Comment by justinas-b** [comment]
I have updated tests @rarkins

**Review by viceice — DISMISSED** [review]

**Comment by justinas-b** [comment]
I have fixed the prettifier issue, however i see that test coverage is failing now. Could you point me to the location where this should be fixed?

```
=============================== Coverage summary ===============================
Statements   : 99.99% ( 32315/32317 )
Branches     : 98.09% ( 10463/10666 )
Functions    : 99.97% ( 3916/3917 )
Lines        : 99.99% ( 31842/31843 )
================================================================================
Jest: "global" coverage threshold for statements (100%) not met: 99.99%
Jest: "global" coverage threshold for lines (100%) not met: 99.99%
Jest: "global" coverage threshold for functions (100%) not met: 99.97%

Test Suites: 561 passed, 561 total
Tests:       11145 passed, 11145 total
Snapshots:   867 passed, 867 total
Time:        902.615 s
Ran all test suites.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Process completed with exit code 1.
```

**Comment by rarkins** [comment]
https://github.com/renovatebot/renovate/blob/36e7cb6e6909f27e26e98538fc42e5fcb048f24b/docs/development/local-development.md?plain=1#L149-L158

**Comment by justinas-b** [comment]
Fixed code coverage

**Review by viceice — APPROVED** [review]

**Comment by Wirone** [comment]
IMHO it's a feature, not a fix, but who cares - the most important is that it's implemented 😅  Thank you @justinas-b 🍻

**Pull request merged**